### PR TITLE
Fix the Composer script handlers

### LINKED ISSUE #1956: Array to string conversion error in ScriptHandler::initializeApplication
**Affected version(s)**

Contao 4.10.0-RC1

**Description**

The execution of the ScriptHandler of the manager-bundle in the Managed Edition is currently broken due to the changes introduced in https://github.com/contao/contao/pull/1917.

The code inspector probably wrongly assumes, that the version of `symfony/process` in use for the ScriptHandler is the one from the Contao dependencies. However, that is not the case within the `composer.phar`, as it brings its own `symfony/process` dependency - and it only uses a lower version, where the first constructor argument of `Process` _must_ be a string, not an array.

Thus we currently get the following output during `composer install/update`:

```
> Contao\ManagerBundle\Composer\ScriptHandler::initializeApplication
Script Contao\ManagerBundle\Composer\ScriptHandler::initializeApplication handling the post-update-cmd event terminated with an exception

  [ErrorException]
  Array to string conversion
```

For reference: the version of `symfony/process` used by the latest stable version of composer (`1.10.9`) is `2.8.52`: https://github.com/composer/composer/blob/83c3250093d5491600a822e176b107a945baf95a/composer.lock#L864

And in version `2.8.52` of `symfony/process` the constructor of `Process` the first parameter needs to be a `string`, not an `array`: https://github.com/symfony/process/blob/c3591a09c78639822b0b290d44edb69bf9f05dc8/Process.php#L127-L137

### PR DESCRIPTION
**Pull request by leofeyer**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #1956
| Docs PR or issue | -

### CODE REVIEW

**Comment by leofeyer** [comment]
However:

```
> Contao\ManagerBundle\Composer\ScriptHandler::initializeApplication
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Symfony\Component\Process\Process::__construct() must be of the type array, string given, called in contao/managed-edition/vendor/contao/contao/manager-bundle/src/Composer/ScriptHandler.php on line 79 and defined in /Users/leofeyer/.composer/vendor/symfony/process/Process.php:140
Stack trace:
#0 contao/managed-edition/vendor/contao/contao/manager-bundle/src/Composer/ScriptHandler.php(79): Symfony\Component\Process\Process->__construct('/usr/local/Cell...')
#1 contao/managed-edition/vendor/contao/contao/manager-bundle/src/Composer/ScriptHandler.php(31): Contao\ManagerBundle\Composer\ScriptHandler::executeCommand(Array, Object(Composer\Script\Event))
#2 /Users/leofeyer/.composer/vendor/composer/composer/src/Composer/EventDispatcher/EventDispatcher.php(305): Contao\ManagerBundle\Composer\ScriptHandler::initializeApplication(Object(Composer\Script\Event))
#3 /Users/leofeyer/.composer/vendor/composer/composer in /Users/leofeyer/.composer/vendor/symfony/process/Process.php on line 140

Fatal error: Uncaught TypeError: Argument 1 passed to Symfony\Component\Process\Process::__construct() must be of the type array, string given, called in contao/managed-edition/vendor/contao/contao/manager-bundle/src/Composer/ScriptHandler.php on line 79 and defined in /Users/leofeyer/.composer/vendor/symfony/process/Process.php:140
Stack trace:
#0 contao/managed-edition/vendor/contao/contao/manager-bundle/src/Composer/ScriptHandler.php(79): Symfony\Component\Process\Process->__construct('/usr/local/Cell...')
#1 contao/managed-edition/vendor/contao/contao/manager-bundle/src/Composer/ScriptHandler.php(31): Contao\ManagerBundle\Composer\ScriptHandler::executeCommand(Array, Object(Composer\Script\Event))
#2 /Users/leofeyer/.composer/vendor/composer/composer/src/Composer/EventDispatcher/EventDispatcher.php(305): Contao\ManagerBundle\Composer\ScriptHandler::initializeApplication(Object(Composer\Script\Event))
#3 /Users/leofeyer/.composer/vendor/composer/composer in /Users/leofeyer/.composer/vendor/symfony/process/Process.php on line 140
```

And PhpStan throws the same error: https://github.com/contao/contao/pull/1964/checks?check_run_id=899092531

@fritzmg So I guess I cannot reproduce the issue?

**Comment by fritzmg** [comment]
> @fritzmg So I guess I cannot reproduce the issue?

Which version of Composer did you use?

**Comment by leofeyer** [comment]
The latest stable: `Composer 1.10.9 2020-07-16 12:57:00`

**Comment by fritzmg** [comment]
Hm, that's odd. Why would it use the _internal_ Symfony Process class only in my case?

**Comment by m-vo** [comment]
The current master is working for me as well. 

@fritzmg Maybe you could set a breakpoint in the script handler and run composer with xdebug enabled to see the call stack.

**Comment by fritzmg** [comment]
@m-vo the call stack is here: https://github.com/contao/contao/issues/1956#issuecomment-662297121

**Comment by fritzmg** [comment]
This is really weird. If it works for you guys _now_, how can it possibly have worked for you _before_ #1917? Because before `4.10.0-RC1` the command was passed as a _string_ to `Process`, not as an _array_.

**Comment by m-vo** [comment]
Looks like this might be a problem that will occur on windows and/or PHP<7.4. :see_no_evil:  

I was running `COMPOSER_ALLOW_XDEBUG=1 php -dxdebug.remote_enable=1 -dxdebug.remote_autostart=1  composer.phar run post-update-cmd` and put a breakpoint at the failing line from your callstack. But it wasn't hit. Turns out there is special handling in place for windows:

```php
if ('\\' === \DIRECTORY_SEPARATOR && $this->enhanceWindowsCompatibility) {
    $commandline = 'cmd /V:ON /E:ON /D /C "(' . $commandline . ')';
    
    foreach ($this->processPipes->getFiles() as $offset => $filename) {
        $commandline .= ' ' . $offset . '>' . ProcessUtils::escapeArgument($filename);
    }
    
    $commandline .= '"';

    if (!isset($this->options['bypass_shell'])) {
        $this->options['bypass_shell'] = true;
    }
} elseif (!$this->useFileHandles && $this->enhanceSigchildCompatibility && $this->isSigchildEnabled()) {
    // ...
}

$this->process = proc_open($commandline, $descriptors, $this->processPipes->pipes, $this->cwd, $this->env, $this->options);
```

So that's why it's failing here: `$commandline = 'cmd /V:ON /E:ON /D /C "(' . $commandline . ')';` if we're passing an array.

Also passing an array as the first argument of `proc_open` (see last line) is only possible since [PHP7.4](https://www.php.net/manual/de/function.proc-open.php) and will e.g. result in the following error on PHP7.3:
```
 [ErrorException]                                           
  proc_open() expects parameter 1 to be string, array given 
```

**Comment by fritzmg** [comment]
@m-vo shouldn't the same problem happen here on Linux? https://github.com/symfony/process/blob/c3591a09c78639822b0b290d44edb69bf9f05dc8/Process.php#L277

Also which `Process.php` gets used in your case actually? The one from the Phar or the one from your project's vendor?

**Comment by m-vo** [comment]
I'm not hitting the `elseif` branch, so at least not in my case.

**Comment by m-vo** [comment]
> Also which Process.php gets used in your case actually? The one from the Phar or the one from your project's vendor?

The one from the phar. (This is actually the problem. I thought composer starts a new process when it executes a script handler but it doesn't. I'll debug it).

**Comment by m-vo** [comment]
In a script handler we cannot use the same components in another version than those composer is using. Or in other words: it will use the already loaded ones from the current context. In our `ScriptHandler` this is e.g. relevant for the `Filesystem` or the `Process` component. We would probably be better off if the script handler was an executable itself.

For reference, see https://stackoverflow.com/questions/52761775/how-to-force-composer-script-to-load-local-class-instead-of-global-class 

Also see: https://github.com/composer/composer/issues/6738#issuecomment-347856605

File: core-bundle/src/Composer/ScriptHandler.php [hunk 1]
```diff
@@ -70,7 +70,8 @@ private static function executeCommand(array $cmd, Event $event): void
             $command[] = $verbose;
         }
 
-        $process = new Process($command);
+        // Composer uses an old version of symfony/console which expects a string (see #1956)
+        $process = new Process(implode(' ', $command));
```
Comment (m-vo) [inline core-bundle/src/Composer/ScriptHandler.php]: We could feature detect the version: see https://github.com/symfony/symfony/commit/8895bc1b5bab9af56fbc4458337f83046347bab5

```suggestion
        // Older versions of symfony/process expect a string (see #1956)
        if (method_exists(Process::class, 'setCommandline')) {
            $command = implode(' ', $command);
        }

        $process = new Process($command);
```

File: manager-bundle/src/Composer/ScriptHandler.php [hunk 2]
```diff
@@ -75,7 +75,8 @@ private static function executeCommand(array $cmd, Event $event, string $env = '
             $command[] = $verbose;
         }
 
-        $process = new Process($command);
+        // Composer uses an old version of symfony/console which expects a string (see #1956)
+        $process = new Process(implode(' ', $command));
```
Comment (m-vo) [inline manager-bundle/src/Composer/ScriptHandler.php]:
```suggestion
        // Older versions of symfony/process expect a string (see #1956)
        if (method_exists(Process::class, 'setCommandline')) {
            $command = implode(' ', $command);
        }

        $process = new Process($command);
```

**Comment by m-vo** [comment]
@leofeyer I could not reproduce your error either by the way. :see_no_evil:

**Comment by fritzmg** [comment]
> We would probably be better off if the script handler was an executable itself.

That sounds good in theory, unfortunately it will lead to problems again on Windows 😁. Reason being: if you have a composer script like this:

```json
{
    "scripts": {
        "unit-tests": [
            "vendor/bin/phpunit --colors=always"
        ]
    }
}
```
like `contao/contao` does for example, it will fail:

```
$ composer unit-tests
> vendor/bin/phpunit --colors=always
'vendor' is not recognized as an internal or external command,
operable program or batch file.
Script vendor/bin/phpunit --colors=always handling the unit-tests event returned with error code 1
```

This is because composer will execute this with Symfony Process which will execute it within `CMD`, even if you executed `composer unit-tests` on the Git Bash for example. And this in turn means that paths with forward slashes won't work ...

**Comment by leofeyer** [comment]
I like @m-vo's `method_exists()` idea. @fritzmg Does it fix the problem for you?

**Comment by m-vo** [comment]
I'd vote for a quick fix where we detect the `symfony/process` version and act accordingly for now. If we manage to run an isolated process for the handlers that works across platforms (might need symfony/composer fixes as well) this should imo be the way to go.

We must be compatible to these versions of `symfony/process`:
https://github.com/composer/composer/blob/1.10/composer.json#L37 

Imo it's enough to check if the `setCommandline` function exists as it was removed in the same commit as the change from string to array args was introduced.

**Comment by fritzmg** [comment]
@leofeyer yes, that fix works for me.

File: core-bundle/src/Composer/ScriptHandler.php [hunk 3]
```diff
@@ -70,8 +70,12 @@ private static function executeCommand(array $cmd, Event $event): void
             $command[] = $verbose;
         }
 
-        // Composer uses an old version of symfony/console which expects a string (see #1956)
-        $process = new Process(implode(' ', $command));
+        // Backwards compatibility with symfony/process <3.3 (see #1964)
```
Comment (m-vo) [inline core-bundle/src/Composer/ScriptHandler.php:73]: Is this 3.3 instead of 2.7 because people would use a `composer.phar` built for PHP7.2?

Comment (leofeyer) [reply core-bundle/src/Composer/ScriptHandler.php:73]: It is `3.3` because any prior version does not handle an array.

Comment (m-vo) [reply core-bundle/src/Composer/ScriptHandler.php:73]: Ah my bad, misread the comment.

**Review by m-vo — APPROVED** [review]

**Review by fritzmg — APPROVED** [review]

File: core-bundle/src/Composer/ScriptHandler.php [hunk 4]
```diff
@@ -70,6 +70,11 @@ private static function executeCommand(array $cmd, Event $event): void
             $command[] = $verbose;
         }
 
+        // Backwards compatibility with symfony/process <3.3 (see #1964)
+        if (method_exists(Process::class, 'setCommandline')) {
+            $command = implode(' ', $command);
```
Comment (ausi) [inline core-bundle/src/Composer/ScriptHandler.php:75]: Isn’t it a bit risky to just implode the command line parts without escaping?

```suggestion
            $command = implode(' ', array_map('escapeshellarg', $command));
```

File: manager-bundle/src/Composer/ScriptHandler.php [hunk 5]
```diff
@@ -75,6 +75,11 @@ private static function executeCommand(array $cmd, Event $event, string $env = '
             $command[] = $verbose;
         }
 
+        // Backwards compatibility with symfony/process <3.3 (see #1964)
+        if (method_exists(Process::class, 'setCommandline')) {
+            $command = implode(' ', $command);
```
Comment (ausi) [inline manager-bundle/src/Composer/ScriptHandler.php:80]:
```suggestion
            $command = implode(' ', array_map('escapeshellarg', $command));
```

File: core-bundle/src/Composer/ScriptHandler.php (hunk 4 above)
Comment (leofeyer) [reply core-bundle/src/Composer/ScriptHandler.php:75]: See #2026

**Pull request merged**